phd-tests/tests/src/boot_order.rs

```diff
@@ -0,0 +1,673 @@
+                [..file_path_list_length as usize],
+        );
+
+        let path_entry = DevicePath::parse_from(&mut device_path_cursor)
+            .expect("can read device path element");
+        let load_path = match path_entry {
+            acpi_root @ DevicePath::Acpi { .. } => {
+                let pci_device =
+                    DevicePath::parse_from(&mut device_path_cursor)
+                        .expect("can read device path element");
+                if !matches!(pci_device, DevicePath::Pci { .. }) {
+                    bail!("expected ACPI Device Path entry to be followed by a PCI Device Path, but was {:?}", pci_device);
+                }
+
+                EfiLoadPath::Device { acpi_root, pci_device }
+            }
+            volume @ DevicePath::FirmwareVolume { .. } => {
+                let file = DevicePath::parse_from(&mut device_path_cursor)
+                    .expect("can read device path element");
+                if !matches!(file, DevicePath::FirmwareFile { .. }) {
+                    bail!("expected Firmware Volume entry to be followed by a Firmware File, but was {:?}", file);
+                }
+
+                EfiLoadPath::FirmwareFile { volume, file }
+            }
+            other => {
+                bail!("unexpected root EFI Load Option path item: {:?}", other);
+            }
+        };
+
+        // Not strictly necessary, but advance `bytes` by the number of bytes we read from
+        // `device_path_cursor`. To callers, this keeps it as if we had just been reading `bytes`
+        // all along.
+        bytes.set_position(bytes.position() + device_path_cursor.position());
+
+        Ok(EfiLoadOption { description, path: load_path })
+    }
+
+    fn pci_device_function(&self) -> (u8, u8) {
+        let EfiLoadPath::Device {
+            pci_device: DevicePath::Pci { device, function },
+            ..
+        } = self.path
+        else {
+            panic!(
+                "expected load path to be an ACPI/PCI pair, but was {:?}",
+                self.path
+            );
+        };
+        (device, function)
+    }
+}
+
+fn unhex(s: &str) -> Vec<u8> {
+    let s = s.replace("\n", "");
+    let mut res = Vec::new();
+    for chunk in s.as_bytes().chunks(2) {
+        assert_eq!(chunk.len(), 2);
+
+        let s = std::str::from_utf8(chunk).expect("valid string");
+
+        let b = u8::from_str_radix(s, 16).expect("can parse");
+
+        res.push(b);
+    }
+    return res;
+}
+
+async fn run_long_command(
+    vm: &phd_framework::TestVm,
+    cmd: &str,
+) -> Result<String, Error> {
+    // Ok, this is a bit whacky: something about the line wrapping for long commands causes
+    // `run_shell_command` to hang instead of ever actually seeing a response prompt.
+    //
+    // I haven't gone and debugged that; instead, chunk the input command up into segments short
+    // enough to not wrap when input, put them all in a file, then run the file.
```

Comment:
@gjcolombo i figure i'll open an issue about this with phd and reference here, but leave this as a workaround for the time being. i couldn't exactly figure out where line wrapping was causing issues but judging from how Alpine and Ubuntu look over serial to my eyes i can certainly understand this being... difficult to automate robustly.

Reply:
My initial wild guess would be that the guest is inserting line control characters when it thinks a line has wrapped, which breaks `run_shell_command`'s logic for determining whether the input command was successfully echoed.

Enabling trace-level logging will make the serial buffering code dump its buffer whenever it checks to see if a wait-for-text request has been satisfied. Verbose as all get out, but should give us some idea why it doesn't think the waited-for text has been seen.

Reply:
(gonna file an issue for this and look into it ~soon)